src/main/ruby/truffleruby/core/truffle/io_operations.rb

```diff
@@ -72,6 +72,15 @@ def self.puts(io, *args)
       nil
     end
 
+    def self.write_transcoding(string, external_encoding)
+      if external_encoding && external_encoding != string.encoding && external_encoding != Encoding::BINARY &&
```

Comment:
The logic used in `StringIO#write` in the original commit was similar to, but different from, the logic in `IO#write`, which I think is leading to some test failures now. For `StringIO` I had to add a check for `str.encoding != Encoding::BINARY` to get the tests to pass.

Reply:
I've pushed an amended commit fixing this up.

Reply:
This is not correct for IO#write, which does raise if the string encoding is BINARY:
```
$ ruby -ve 'File.open("foo.txt", "w", external_encoding: Encoding::UTF_16BE) { |f| f.write "été".b }'
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [x86_64-linux]
-e:1:in `write': "\\xC3" from ASCII-8BIT to UTF-8 (Encoding::UndefinedConversionError)
```

But StringIO is again inconsistent and noops there (which is arguably a CRuby/StringIO bug):
```
ruby -rstringio -ve 's=StringIO.new.set_encoding(Encoding::UTF_16BE); s.write "été".b; p s.string.bytes'
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [x86_64-linux]
[195, 169, 116, 195, 169]
```